### PR TITLE
Shipping rates networking

### LINKED ISSUE #12286: Get label rates
This task is for adding network support for getting label rates using the latest plugin endpoint `/wcshipping/v1/label/rate`

Here is the endpoint's documentation
https://github.com/woocommerce/woocommerce-shipping/blob/trunk/docs/api/label-rate.md

### PR DESCRIPTION
**Pull request by atorresveiga**
Closes: #12286 

### Description
This PR adds the network layer to connect shipping rates to the shipping labels' main flow. Besides its main goal of fetching shipping rates, I also reorganized the shipping rate classes into new packages (network, data source, and domain).
I also included a small improvement in this PR: sorting rates locally and removing the extra API request.

⚠️ I will add mapper classes unit tests in this PR https://github.com/woocommerce/woocommerce-android/pull/13138

### Testing information
1. Open the orders section
2. Select an order that is eligible to create shipping labels (processing state)
3. Tap on create shipping label
4. Select a package
5. Check that shipping rates are fetched from the API
6. Change the sorting order
7. Check that the shipping rates order is changed locally

### The tests that have been performed
1. Testing the UI works as expected
2. Testing the API response matches the rates displayed in the UI

### Images/gif

https://github.com/user-attachments/assets/5b6e6ebf-ce37-47db-a7b1-e19588ed590f

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt [hunk 1]
```diff
@@ -46,60 +51,97 @@ class WooShippingLabelCreationViewModel @Inject constructor(
         originCountry = "US"
     )
 
+    private val order = MutableStateFlow<Order?>(null)
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt]: I decided to keep some variables as mutable flow so the UI became more reactive

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt [hunk 2]
```diff
@@ -46,60 +51,97 @@ class WooShippingLabelCreationViewModel @Inject constructor(
...
+    private val storeOptions = MutableStateFlow<StoreOptionsModel?>(null)
     private val shippableItems = MutableStateFlow<List<ShippableItemModel>>(emptyList())
-    private val storeOptions = MutableStateFlow(mockStoreOptions)
+
     private val selectedRatesSortOrder = MutableStateFlow(ShippingSortOption.FASTEST)
     private val refreshShippingRates = MutableSharedFlow<Unit>()
 
     private val packageSelection = MutableStateFlow<PackageSelectionState>(NotSelected)
 
+    private val cheapestComparator = Comparator<ShippingRateUI> { r1, r2 -> r1.price.compareTo(r2.price) }
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt]: Moved the sorting logic to the view model and avoid unnecessary API calls on sorting

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/rates/datasource/WooShippingRatesDatasourceMapper.kt [hunk 3]
```diff
@@ -0,0 +1,71 @@
+package com.woocommerce.android.ui.orders.wooshippinglabels.rates.datasource
+
+import com.woocommerce.android.ui.orders.wooshippinglabels.models.WooShippingCarrier
+import com.woocommerce.android.ui.orders.wooshippinglabels.rates.datasource.WooShippingRateModel.Option
+import com.woocommerce.android.ui.orders.wooshippinglabels.rates.networking.WooShippingRatesDTO
+import javax.inject.Inject
+
+class WooShippingRatesDatasourceMapper @Inject constructor() {
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/rates/datasource/WooShippingRatesDatasourceMapper.kt:8]: I made some extra changes to the models in the next PR tackling selecting shipping labels. I will cover the mappers unit test on that PR

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/rates/domain/GetShippingRates.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import com.woocommerce.android.model.Address
+import com.woocommerce.android.ui.orders.wooshippinglabels.models.OriginShippingAddress
+import com.woocommerce.android.ui.orders.wooshippinglabels.packages.ui.PackageData
+import com.woocommerce.android.ui.orders.wooshippinglabels.rates.datasource.WooShippingRatesRepository
+import com.woocommerce.android.ui.orders.wooshippinglabels.rates.ui.CarrierUI
+import com.woocommerce.android.ui.orders.wooshippinglabels.rates.ui.ShippingRateUI
+import com.woocommerce.android.ui.orders.wooshippinglabels.rates.ui.ShippingSortOption
+import javax.inject.Inject
+
+class GetShippingRates @Inject constructor(
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/rates/domain/GetShippingRates.kt:11]: I will cover this use case unit tests in the next PR

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt (hunk 1 above)
Comment (ThomazFB) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt]: I don't think this needs to be adjusted now, but I'm concerned about the savedState handling. We're not associating the ViewState in any way to the ViewModel saved state and that might affect the ability to recover data after the UI is destroyed. We might want to at least configure the `viewState` instance to be initialized from the `savedState.getStateFlow` call.

**Review by ThomazFB — APPROVED** [review]
Great work with the overall structure and architecture approach @atorresveiga. And thanks for adding the `id` field mapping from the DTO all the way through the PackageData! 

The unit test coverage needs an improvement but since you already described that it will be introduced in the next PR, I'm approving this PR. Just one thing: there are some CI errors with the unit tests and detekt that you will need to adjust.

**Pull request merged**